### PR TITLE
fix(ProgressReport):  Replace progress reporting

### LINKED ISSUE #519: Progress bar appending instead of replacing values inline (Macos 10.14.5)
**Describe the bug**
Currently when running dotnet stryker, the reporter does not replace values inline in the terminal. It appends to it. This makes it look like the following:

**Logs**
[10:54:19 INF] Started initial build using dotnet build
[10:54:24 INF] Initial build successful
[10:54:24 INF] Using testrunner DotnetTest
[10:54:24 INF] Initial testrun started
[10:54:29 INF] Total number of tests found in initial test run: 56
[10:54:29 INF] Initial testrun successful in 4518 ms
[10:54:29 INF] Using 36777 ms as testrun timeout
[10:54:32 INF] 53 mutants ready for test

53 mutants have been created. Each mutant will now be tested, this could take a while.

Tests progress | ---------- | 0 / 53 | 0 % | NA |
Killed : 0
Survived: 0
Timeout : 0

Tests progress | ---------- | 1 / 53 | 1 % | ~5m 58s |

Killed : 1

Tests progress | ---------- | 2 / 53 | 3 % | ~2m 56s |

Killed : 2

Tests progress | ---------- | 3 / 53 | 5 % | ~1m 55s |

Survived: 1

Tests progress | ---------- | 4 / 53 | 7 % | ~1m 25s |

**Expected behavior**
I expect the output to be replaced, and thus having an animating progress bar.

**Desktop:**
 - OS: Macos
-  Terminal: Iterm2, and integrated terminal. Using the ZSH framework.
 - Type of project: core
 - Framework Version: core 2.2
 - Stryker Version:  beta 0.11.0

### PR DESCRIPTION
**Pull request by eNeRGy164**
Implement ShellProgressBar and make MutantsResultReporter obsolete

This is a WIP, to first get some feedback.
Added Unit testing.

Closes #519

### CODE REVIEW

File: src/Stryker.Core/Stryker.Core/Reporters/Progress/StopWatchProvider.cs [hunk 1]
```diff
@@ -5,17 +5,22 @@ namespace Stryker.Core.Reporters.Progress
...
         private Stopwatch _watch;
 
         public void Start()
         {
-            _watch = new Stopwatch();
-           _watch.Start();
+            _watch = Stopwatch.StartNew();
+        }
+
+        public void Stop()
```
Comment (Mobrockers) [inline src/Stryker.Core/Stryker.Core/Reporters/Progress/StopWatchProvider.cs:21]: 🙈 nice

**Comment by Mobrockers** [comment]
Looks good to me 🐝

File: src/Stryker.Core/Stryker.Core/Reporters/Progress/StopWatchProvider.cs (hunk 1 above)
Comment (eNeRGy164) [reply src/Stryker.Core/Stryker.Core/Reporters/Progress/StopWatchProvider.cs:21]: I thought so :P

**Comment by Mobrockers** [comment]
Mhm after testing it looks like the fancy progress bar does not work correctly in the visual studio debug console

image

In git bash it fails completely

image

It does work in CMD:

image


Git bash and the visual studio debug console are probably not the main shells used to run stryker but it suggests to me that the used library does not support all shells so a bit more error handling could not hurt.

Especially the issue in git bash is concerning because that is the same issue we noticed on macOSX before and is why we disabled replacing values when Console.CursorTop caused an exception:

https://github.com/stryker-mutator/stryker-net/blob/b05a5f907f9db4ff0f0b4502a15bc675cb3a993f/src/Stryker.Core/Stryker.Core/Reporters/Progress/ConsoleOneLineLogger.cs#L62-L74

**Comment by eNeRGy164** [comment]
https://github.com/Mpdreamz/shellprogressbar/issues/22
Some of these seem to be there for a while.

**Comment by Mobrockers** [comment]
Does anyone know someone with a mac that could test this pr? 😅

**Comment by richardwerkman** [comment]
@vngrp has a mac, maybe he could take a look?

**Comment by eNeRGy164** [comment]
I have no environment to recreate these issues.

Works for me:
* Windows PowerShell 5.1
* PowerShell Core 7
* Windows cmd
* PowerShell terminal in VSCode
  image
* Visual Studio Debug Conole
  image

**Comment by eNeRGy164** [comment]
> _Mhm after testing it looks like the fancy progress bar does not work correctly in the visual studio debug console_
> 
> This happens only when debug logging is turned on

Ah missed that remark.
Yes, you can't have multiple processes doing console output at the same time if one is messing with the positions, like in this case the progressbar. That is also why I had to merge the mutant reporting and the progressbar in one output.

**Comment by Mobrockers** [comment]
> > _Mhm after testing it looks like the fancy progress bar does not work correctly in the visual studio debug console_
> > This happens only when debug logging is turned on
> 
> Ah missed that remark.
> Yes, you can't have multiple processes doing console output at the same time if one is messing with the positions, like in this case the progressbar. That is also why I had to merge the mutant reporting and the progressbar in one output.

Unfortunately that seems like a regression compared to the current implementation (on some terminals).
We're going to do more testing to decide what to do with this pr, as it stands it looks like we're gonna have to decide between regressions on the one hand or other issues on the other hand.

Sorry for the wait.

**Comment by eNeRGy164** [comment]
No problem, it's your project after all.
I just can't reproduce the errors, so it's hard to fix them myself.

**Comment by eNeRGy164** [comment]
For showing addition logging we could reroute to the WriteLine function mentioned here: https://github.com/Mpdreamz/shellprogressbar/issues/31#issuecomment-523022264

**Pull request merged**